From: Oswald Buddenhagen <oswald.buddenhagen@theqtcompany.com>
Date: Tue, 05 Jul 2016 16:07:14 +0000
Subject: fix raw data leak in $$absolute_path()
X-Git-Url: http://quickgit.kde.org/?p=qt%2Fqtbase.git&a=commitdiff&h=3c8134958c66f40bb86588aa91b83bf58b5de0c9
---
fix raw data leak in $$absolute_path()

when the file name is empty, the path will be returned verbatim. this
must be considered when constructing the return value.

Task-number: QTBUG-54550
Change-Id: Ie108ed52275e66a154ef63bd6f7193f55b3e0454
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
---


--- a/qmake/library/qmakebuiltins.cpp
+++ b/qmake/library/qmakebuiltins.cpp
@@ -1027,7 +1027,11 @@
             QString rstr = QDir::cleanPath(
                     QDir(args.count() > 1 ? args.at(1).toQString(m_tmp2) : currentDirectory())
                     .absoluteFilePath(args.at(0).toQString(m_tmp1)));
-            ret << (rstr.isSharedWith(m_tmp1) ? args.at(0) : ProString(rstr).setSource(args.at(0)));
+            ret << (rstr.isSharedWith(m_tmp1)
+                        ? args.at(0)
+                        : args.count() > 1 && rstr.isSharedWith(m_tmp2)
+                            ? args.at(1)
+                            : ProString(rstr).setSource(args.at(0)));
         }
         break;
     case E_RELATIVE_PATH:

--- a/tests/auto/tools/qmakelib/evaltest.cpp
+++ b/tests/auto/tools/qmakelib/evaltest.cpp
@@ -2481,6 +2481,14 @@
                "Project MESSAGE: assign split separate: word this is a test done\n"
                "Project MESSAGE: assign split joined: word: this is a test:done\n"
                "Project MESSAGE: assign split quoted: word   this   is a     test done"
+            << true;
+
+    // Raw data leak with empty file name. Verify with Valgrind or asan.
+    QTest::newRow("QTBUG-54550")
+            << "FULL = /there/is\n"
+               "VAR = $$absolute_path(, $$FULL/nothing/here/really)"
+            << "VAR = /there/is/nothing/here/really"
+            << ""
             << true;
 }
 

